src/main/java/beam/physsim/cch/RoutingFrameworkTravelTimeCalculator.scala

```diff
@@ -0,0 +1,230 @@
+                      speedFoundForAtLeastOneHour = true
+                      travelTime / linksInWay.size
+
+                    case None =>
+                      link.getLength / link.getFreespeed
+                  }
+                }
+
+                if (!speedFoundForAtLeastOneHour)
+                  linksFailedToResolve = linksFailedToResolve + 1
+
+                linkTravelTimeByHour
+
+              case None =>
+                linksFailedToResolve = linksFailedToResolve + 1
+                Array.fill(maxHour) { link.getLength / link.getFreespeed }
+            }
+            link.getId.toString -> travelTimeByHour
+          }
+      }
+
+    logger.info("Total links: {}, failed to assign travel time: {}", totalNumberOfLinks, linksFailedToResolve)
+
+    logger.info("Created travel times in {} ms", System.currentTimeMillis - startTime)
+
+    linkId2TravelTimeByHour.asJava
+  }
+
+  private def generateHour2Events(
+    pathTraversalEvents: util.Collection[PathTraversalEvent]
+  ): Map[Int, List[TravelInfo]] = {
+    val preliminaryHour2Events: Map[Int, List[PathTraversalEvent]] =
+      pathTraversalEvents.toStream
+        .map(x => x.departureTime / 3600 -> x)
+        .groupBy(_._1)
+        .mapValues(_.map(_._2).toList)
+
+    val hours2EventsMutable = mutable.HashMap[Int, mutable.ArrayBuffer[TravelInfo]]()
+
+    preliminaryHour2Events.foreach {
+      case (hour, events) =>
+        events.foreach { event =>
+          val currentLinkIds = mutable.ArrayBuffer[Int]()
+          var currentHour = hour
+
+          var currentTime = event.departureTime.toDouble
+          event.linkTravelTime.zip(event.linkIds).foreach {
+            case (travelTime, linkId) =>
+              currentLinkIds += linkId
+              currentTime = currentTime + travelTime
+              val possiblyNextHour = currentTime.toInt / 3600
+
+              if (possiblyNextHour != currentHour) {
+                val currentHourEvents =
+                  hours2EventsMutable.getOrElseUpdate(currentHour, new mutable.ArrayBuffer[TravelInfo]())
+                currentHourEvents += TravelInfo(currentLinkIds.toIndexedSeq)
+                currentLinkIds.clear()
+
+                currentHour = possiblyNextHour
+              }
+          }
+          if (currentLinkIds.nonEmpty) {
+            val currentHourEvents =
+              hours2EventsMutable.getOrElseUpdate(currentHour, new mutable.ArrayBuffer[TravelInfo]())
+            currentHourEvents += TravelInfo(currentLinkIds.toIndexedSeq)
+          }
+        }
+    }
+
+    hours2EventsMutable.mapValues(_.toList).toMap
+  }
+
+  private def linkWayId(link: Link): Option[Long] =
+    Option(link.getAttributes.getAttribute("origid"))
+      .map(_.toString.toLong)
+
+  private def getClosestVertexId(
+    coordinate: Coordinate
+  ): Long = {
+    val vertexes =
+      coordKey2Coordinates.getOrElse(coordKey(coordinate), Nil)
+
+    if (vertexes.nonEmpty) {
```

Comment:
Should it be replaced by QuadTree or this fallback is very rare?

Reply:
replaced with querying from KDtree with gradually increasing envelope